src/core/device.c

```diff
@@ -114,12 +160,128 @@ static void device_done(Unit *u) {
         assert(d);
 
         device_unset_sysfs(d);
+        free(d->devname);
         d->wants_property = strv_free(d->wants_property);
 }
 
+static int root_cgroup_apply_block_device_io_cost_qos(Unit *root_slice_unit, Device *d) {
+        char buf[DECIMAL_STR_MAX(dev_t)*2+9+(6+DECIMAL_STR_MAX(uint64_t)+1)*10+(6+DECIMAL_STR_MAX(loadavg_t)+1)*2];
```

Comment:
I'd pull this out into a `define` and add documentation on what all these numbers mean for the next person looking at this code.

Reply:
These numbers are based on the expected contents of the string we will write to the /sys/fs/cgroup/io.cost.qos. It is specific to io.cost.qos (or io.cost.model). This is based off the way similar functions are implemented. See `cgroup_apply_unified_cpu_weight` and similar functions in cgroup.c.